scripts/clearProperties.ts

```diff
@@ -0,0 +1,89 @@
+
+    return items.filter(t => t.propertyName.endsWith(domain)).filter(t => t.propertyName !== testDomain.hostname);
+}
+
+const checkDNSRecordExists = async (domain: string) => {
+    const response = await fetch(`https://api.cloudflare.com/client/v4/zones/${process.env.CF_ZONE_ID}/dns_records?name=${domain}`, {
+        headers: {
+            Authorization: `Bearer ${process.env.CF_AUTH_TOKEN}`,
+        }
+    })
+    const body = await response.json();
+
+    return body.result_info.total_count > 0;
+}
+
+const getActiveVersions = async (propertyId: string) => {
+    const {body: {activations: {items}}} = await akamaiRequest({
+        path: `/papi/v1/properties/${propertyId}/activations?contractId=${process.env.AK_CONTRACT_ID}&groupId=${process.env.AK_GROUP_ID}`,
+        method: 'GET',
+    });
+
+    return items.filter(t => t.status === 'ACTIVE' && t.activationType === 'ACTIVATE');
+}
+
+const deActivateVersion = async (propertyId: string, {network, version}: {
+    network: string,
+    version: number
+}) => {
+    return akamaiRequest({
+        path: `/papi/v1/properties/${propertyId}/activations?contractId=${process.env.AK_CONTRACT_ID}&groupId=${process.env.AK_GROUP_ID}`,
+        method: 'POST',
+        body: JSON.stringify({
+            network,
+            propertyVersion: version,
+            activationType: "DEACTIVATE",
+            notifyEmails: ['support+akamai@fingerprint.com'],
+            acknowledgeAllWarnings: true,
+        })
+    });
+}
+
+const deleteProperty = async (propertyId: string) => {
+    return akamaiRequest({
+        path: `/papi/v1/properties/${propertyId}?contractId=${process.env.AK_CONTRACT_ID}&groupId=${process.env.AK_GROUP_ID}`,
+        method: 'DELETE',
+    });
+};
+const handler = async () => {
+    const properties = await filterProperties();
+    for (const property of properties) {
+        const dnsRecordExists = await checkDNSRecordExists(property.propertyName);
+        if (!dnsRecordExists) {
+            const activations = await getActiveVersions(property.propertyId);
+            const versionsToDeactivate = activations.map(t => ({
+                version: t.propertyVersion,
+                network: t.network,
+                activationId: t.activationId
+            }));
+            for (const activation of versionsToDeactivate) {
+                try {
+                    await deActivateVersion(property.propertyId, {
+                        network: activation.network,
+                        version: activation.version
+                    });
+                    console.log('SUCCESS:DEACTIVATE', `property(${property.propertyId}) version(${activation.version}) network(${activation.network})`);
+                } catch (_) {
+                }
+            }
+            try {
+                await deleteProperty(property.propertyId);
+                console.log('SUCCESS:DELETE', `property(${property.propertyId})`);
+            } catch (_) {
```

Comment:
I don't like a few things here:
1. We try to delete the property even when there are active versions. The property will not be deleted if all activations are deactivated, so we shouldn't try to delete the property.
2. `getActiveVersions` only returns the activations that are done, right? If we are going to re-write the logic here, the new logic should check if there are versions that are pending to be activated/deactivated, if yes, it shouldn't try to delete the property since it will surely fail anyway.
3. We don't log any errors on deactivation and deleting the property. We won't know if there is something wrong and we should fix it.

Reply:
1. It's just increasing our cost, we can simply just send a delete request and Akamai can tell us "Not right now".
2. getActiveVersions not only checking for completed activations, it checks all the activations, try to deactivate/cancel.
3. Logging errors may or may not contain sensitive information about Fingerprint's Akamai Account. This why we don't share logs on public repository.

Reply:
We discussed and concluded this in a huddle.